src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -0,0 +1,266 @@
+    beast::Journal& j)
+{
+    static boost::regex const re (
+        "^"                         // start of line
+        "(?:\\s*)"                  // whitespace (optional)
+        "([a-zA-Z0-9]+)"            // Node identity
+        "(?:\\s*)"                  // whitespace (optional)
+        "(.*)"                      // comment (optional)
+    );
+
+    std::size_t count = 0;
+
+    for (auto const& n : entries)
+    {
+        boost::smatch match;
+
+        if (!boost::regex_match (n, match, re))
+        {
+            JLOG (j.error) <<
+                "Malformed entry: '" << n << "'";
+            continue;
+        }
+
+        auto const id = parseBase58<PublicKey>(
+            TokenType::TOKEN_NODE_PUBLIC, match[1]);
+
+        if (!id)
+        {
+            JLOG (j.error) <<
+                "Invalid node identity: " << match[1];
+            continue;
+        }
+
+        auto const ra = RippleAddress::createNodePublic (match[1]);
+
+        if (vl->member (ra))
+        {
+            JLOG (j.warning) <<
+                "Duplicate node identity: " << match[1];
+            continue;
+        }
+
+        if (vl->insertPermanentKey(ra, trim_whitespace (match[2])))
+            count++;
+    }
+
+    return count;
+}
+
+//------------------------------------------------------------------------------
+std::unique_ptr<ValidatorList>
+make_ValidatorList (Config const& config, beast::Journal j)
+{
+    auto validatorList = std::make_unique<ValidatorList> (j);
+
+    std::size_t const count = loadValidators (
+        validatorList.get(),
+        config.validators,
+        j);
+
+    JLOG (j.debug) <<
+        "Loaded " << count <<
+        " entries from the configuration file";
+
+    auto file = config.VALIDATORS_FILE;
+
+    if (!file.empty ())
+    {
+        if (boost::filesystem::exists (file))
+        {
+            JLOG (j.error) <<
+                "Validators file " << file <<
+                " does not exist";
+        }
+        else if (boost::filesystem::is_regular_file (file))
+        {
+            JLOG (j.error) <<
+                "Validators file " << file <<
+                " is not a regular file";
+        }
+        else
+        {
+            std::ifstream ifsDefault (
+                file.native ().c_str (),
+                std::ios::in);
+
+            std::string data;
+
+            data.assign (
+                std::istreambuf_iterator<char> (ifsDefault),
+                std::istreambuf_iterator<char> ());
+
+            auto secValidators = parseIniFile (data, true);
+
+            auto entries = getIniFileSection (
+                secValidators,
+                SECTION_VALIDATORS);
+
+            if (!entries)
+            {
+                JLOG (j.error) <<
+                    "Validators file " << file <<
+                    " does not contain a [" SECTION_VALIDATORS <<
+                    "] section";
+                return false;
```

Comment:
The callers don't check for an empty pointer, so returning `nullptr` would require additional work in Application.cpp